File: 1129-shortest-path-with-alternating-colors/1129-shortest-path-with-alternating-colors.cpp
typedef long long LL;
const int OO = 10000000; // A big value expressing infinity
const int RED = 0;
const int BLUE = 1;

struct edge
{
    int to;
    int color;
};

typedef vector<vector<edge>> GRAPH;

void add_directed_edge(GRAPH &graph, int from, int to, int color)
{
    graph[from].push_back({to, color});
}

vector<int> bfs(GRAPH &adjList, int start)
{
    int nodes = adjList.size();
    vector<int> answer(nodes, -1);
    vector<vector<int>> len(nodes, vector<int>(2, OO)); // N*2 grid
    queue<edge> q;

    q.push({start, RED});
    q.push({start, BLUE});
    len[start][RED] = len[start][BLUE] = 0;
    answer[start] = 0;

    for (int level = 0, sz = q.size(); !q.empty(); ++level, sz = q.size())
    {
        while (sz--)
        {
            int cur = q.front().to, color = q.front().color;
            q.pop();
            for (edge e : adjList[cur])
                if (color != e.color && len[e.to][e.color] == OO)
                {
                    q.push(e);
                    len[e.to][e.color] = level + 1;
                    if (answer[e.to] == -1) // we can reach the node from 2 colors. Keep the first only
                        answer[e.to] = level + 1;
                }
        }
    }
    return answer;
}

class Solution
{
public:
    vector<int> shortestAlternatingPaths(int n, vector<vector<int>> &red_edges, vector<vector<int>> &blue_edges)
    {
        GRAPH adjList(n);

        for (auto &e : red_edges)
            add_directed_edge(adjList, e[0], e[1], RED);

        for (auto &e : blue_edges)
            add_directed_edge(adjList, e[0], e[1], BLUE);

        return bfs(adjList, 0);
    }
};

File: 1129-shortest-path-with-alternating-colors/1129-shortest-path-with-alternating-colors.java
import java.util.*;

class Solution {
    public static int[] shortestAlternatingPaths(int n, int[][] redEdges, int[][] blueEdges) {
        int[] res = new int[n];
        HashMap<Integer, List<Edge>> graph = new HashMap<>();
        for (int i = 0; i < res.length; i++) {
            res[i] = -1;
        }
        for (var e : redEdges) {
            if (!graph.containsKey(e[0])) {
                List<Edge> temp = new ArrayList();
                temp.add(new Edge(e[1], 'r'));
                graph.put(e[0], temp);

            } else {
                List<Edge> temp = graph.get(e[0]);
                temp.add(new Edge(e[1], 'r'));
                graph.put(e[0], temp);
            }
        }
        for (var e : blueEdges) {
            if (!graph.containsKey(e[0])) {
                List<Edge> temp = new ArrayList();
                temp.add(new Edge(e[1], 'b'));
                graph.put(e[0], temp);

            } else {
                List<Edge> temp = graph.get(e[0]);
                temp.add(new Edge(e[1], 'b'));
                graph.put(e[0], temp);
            }
        }

        char currentColor = 'x';
        Queue<Edge> queue = new LinkedList<>();
        queue.add(new Edge(0, 'x'));
        HashSet<Edge> vis = new HashSet<>();
        vis.add(new Edge(0, 'x'));
        res[0] = 0;
        for (int level = 0, sz = queue.size(); !queue.isEmpty(); level++, sz = queue.size()) {
            while (sz-- > 0) {
                Edge edge = queue.poll();
                currentColor = edge.color;
                for (Edge e : graph.getOrDefault(edge.to, new ArrayList<>())) {
                    if (currentColor == e.color || vis.contains(e)) {
                        continue;
                    }
                    queue.add(e);
                    vis.add(e);
                    if (res[e.to] == -1)
                        res[e.to] = level + 1;
                }

            }

        }

        return res;
    }


    static class Edge {
        int to;
        char color;

        public int getTo() {
            return to;
        }

        public void setTo(int to) {
            this.to = to;
        }

        public char getColor() {
            return color;
        }

        public void setColor(char color) {
            this.color = color;
        }

        public Edge(int to, char color) {
            this.to = to;
            this.color = color;
        }
    }

    public static void main(String[] args) {
        shortestAlternatingPaths(3, new int[][]{{0, 1}, {1, 2}}, new int[][]{});
    }
}